api/envoy/config/filter/network/thrift_proxy/v2alpha1/route.proto

```diff
@@ -66,8 +70,42 @@ message RouteMatch {
   repeated envoy.api.v2.route.HeaderMatcher headers = 4;
 }
 
-// [#comment:next free field: 2]
+// [#comment:next free field: 3]
 message RouteAction {
-  // Indicates the upstream cluster to which the request should be routed.
-  string cluster = 1 [(validate.rules).string.min_bytes = 1];
+  oneof cluster_specifier {
+    option (validate.required) = true;
+
+    // Indicates a single upstream cluster to which the request should be routed
+    // to.
+    string cluster = 1 [(validate.rules).string.min_bytes = 1];
+
+    // Multiple upstream clusters can be specified for a given route. The
+    // request is routed to one of the upstream clusters based on weights
+    // assigned to each cluster.
+    WeightedCluster weighted_clusters = 2;
+  }
+}
+
+// Allows for specification of multiple upstream clusters along with weights that indicate the
+// percentage of traffic to be forwarded to each cluster. The router selects an upstream cluster
+// based on these weights.
+message WeightedCluster {
+  message ClusterWeight {
+    // Name of the upstream cluster.
+    string name = 1 [(validate.rules).string.min_bytes = 1];
+
+    // An integer between 0 and :ref:`total_weight
+    // <envoy_api_field_config.filter.network.thrift_proxy.v2alpha1.WeightedCluster.total_weight>`.
+    // When a request matches the route, the choice of an upstream cluster is determined by its
+    // weight. The sum of weights across all entries in the clusters array must add up to the
+    // total weight.
+    google.protobuf.UInt32Value weight = 2;
+  }
+
+  // Specifies one or more upstream clusters associated with the route.
+  repeated ClusterWeight clusters = 1 [(validate.rules).repeated .min_items = 1];
+
+  // Specifies the total weight across all clusters. The sum of all cluster weights must equal this
+  // value, which must be greater than 0. Defaults to 100.
+  google.protobuf.UInt32Value total_weight = 3 [(validate.rules).uint32.gte = 1];
```

Comment:
Why is it not the total weight of the cluster as total_weight, but what is the need to specify a totoal weight?

Reply:
This matches the HTTP router's version of the same feature. The history is that the original implementation required cluster weights to sum to 100, which forced weighted cluster selections to be specified in 1% increments. Because the weights can also be modified via the runtime parameter system, there was specific logic in place to handle the case where the runtime-modified weights did not sum to 100. To preserve that override logic for finer granularity weights, we needed to know that total weight independently of the sum of the current cluster weights, so we added a total_weight field.

Reply:
Having said that, I don't think we'll end up supporting runtime manipulation of these weights. The runtime values was created before API v2 and the ability to transmit configuration changes to a running Envoy. Perhaps we should remove total_weight and compute it from the cluster weights at configuration time.

Reply:
Oh interesting -- I didn't realize runtime values were meant as an early solution to dynamic config changes. I was under the impression it was still a relevant feature to support and wanted to leave us space to do that. Can remove.